transports/tcp/src/lib.rs

```diff
@@ -90,297 +96,198 @@ impl $tcp_config {
-                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
-
-            let local_addr = listener.local_addr()?;
-            let port = local_addr.port();
-
-            // Determine all our listen addresses which is either a single local IP address
-            // or (if a wildcard IP address was used) the addresses of all our interfaces,
-            // as reported by `get_if_addrs`.
-            let addrs =
-                if socket_addr.ip().is_unspecified() {
-                    let addrs = host_addresses(port)?;
-                    debug!("Listening on {:?}", addrs.iter().map(|(_, _, ma)| ma).collect::<Vec<_>>());
-                    Addresses::Many(addrs)
-                } else {
-                    let ma = ip_to_multiaddr(local_addr.ip(), port);
-                    debug!("Listening on {:?}", ma);
-                    Addresses::One(ma)
-                };
-
-            // Generate `NewAddress` events for each new `Multiaddr`.
-            let pending = match addrs {
-                Addresses::One(ref ma) => {
-                    let event = ListenerEvent::NewAddress(ma.clone());
-                    let mut list = VecDeque::new();
-                    list.push_back(Ok(event));
-                    list
-                }
-                Addresses::Many(ref aa) => {
-                    aa.iter()
-                        .map(|(_, _, ma)| ma)
-                        .cloned()
-                        .map(ListenerEvent::NewAddress)
-                        .map(Result::Ok)
-                        .collect::<VecDeque<_>>()
-                }
-            };
+    fn create_socket(&self, socket_addr: &SocketAddr) -> io::Result<Socket> {
+        let domain = if socket_addr.is_ipv4() {
+            Domain::ipv4()
+        } else {
+            Domain::ipv6()
+        };
+        let socket = Socket::new(domain, Type::stream(), Some(socket2::Protocol::tcp()))?;
+        if socket_addr.is_ipv6() {
+            socket.set_only_v6(true)?;
+        }
+        if let Some(ttl) = self.ttl {
+            socket.set_ttl(ttl)?;
+        }
+        if let Some(nodelay) = self.nodelay {
+            socket.set_nodelay(nodelay)?;
+        }
+        socket.set_reuse_address(true)?;
+        #[cfg(unix)]
+        if self.port_reuse {
+            socket.set_reuse_port(true)?;
+        }
+        Ok(socket)
+    }
 
-            let listen_stream = $tcp_listen_stream {
-                stream: listener,
-                pause: None,
-                pause_duration: cfg.sleep_on_error,
-                port,
-                addrs,
-                pending,
-                config: cfg
-            };
+    fn do_listen(self, socket_addr: SocketAddr) -> io::Result<TcpListenStream> {
+        let socket = self.create_socket(&socket_addr)?;
+        socket.bind(&socket_addr.into())?;
+        socket.listen(self.backlog as _)?;
+        let listener = Async::new(socket.into_tcp_listener())?;
```

Comment:
It is required because futures from async-net take &self, making them impossible to box without unsafe code. Also I don't really see why another library is required, when async-io is so easy to use.